### ISSUE TITLE
Using ds.join twice with masked values / and inner join

### ISSUE THREAD

**Issue by jhjhagen** [vaexio/vaex#96]
When joining ds1 with ds3 (right join to get ds31),
ds31 = ds1.join(ds3, left_on='a', right_on='c', how='right'):

\# | c | z | a | x
-- | -- | -- | -- | --
0 | 'a' | 1 | 'a' | 1
1 | 'b' | 8 | 'b' | 2
2 | 'e' | 27 | masked | masked
3 | 'f' | 64 | masked | masked

for ds1:

\# | a | x
-- | -- | --
0 | 'a' | 1
1 | 'b' | 2
2 | 'c' | 3

for ds3:

\# | c | z | a | x
-- | -- | -- | -- | --
0 | 'a' | 1 | 'a' | 1
1 | 'b' | 8 | 'b' | 2
2 | 'e' | 27 | masked | masked
3 | 'f' | 64 | masked | masked

Now I want to join (left) ds31 with ds4:
ds314 = ds31.join(ds4, left_on='a', right_on='d', how='left', inplace='False')

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-178-7b64de6bfa10> in <module>()
----> 1 ds314 = ds31.join(ds4, left_on='a', right_on='d', how='left', inplace='False')

/net/gaia/data/users/breddels/python/anaconda3/lib/python3.6/site-packages/vaex/dataset.py in join(self, other, on, left_on, right_on, lsuffix, rsuffix, how, inplace)
   5168             right_values = right.evaluate(right_on)
   5169             # maps from the left_values to row #
-> 5170             index_left = dict(zip(left_values, range(N)))
   5171             # idem for right
   5172             index_other = dict(zip(right_values, range(N_other)))

TypeError: unhashable type: 'MaskedConstant'

for ds4:

\# | d | z4
-- | -- | --
0 | 'a' | 1
1 | 'b' | 16
2 | 'c' | 81
3 | 'd' | 256
4 | 'e' | 625

In other words, vaex tries to find matches in ds4 for the masked rows of ds31.
I would like to have a resulting Table in which all ds4 is matched to ds31. Values for ds314 should be masked (or Nan) if the corresponding rows in ds31 are masked.

The same error occurs when:
ds431 = ds4.join(ds31, left_on='d', right_on='a', how='left', inplace='False')

So: **ds.join can't handle 'masked' values.**

In addition: a feature for **inner join** would be useful too.

**Comment by maartenbreddels** [comment]
Thanks @jhjhagen for the great bug reports, I can confirm this is still an issue.
